### PR TITLE
Add Password grant type to OAuth HTTP endpoints

### PR DESCRIPTION
**Pull request by SanojPunchihewa**
## Purpose

Add Password grant type to OAuth HTTP endpoints.

## User Stories

A sample Password grant type config is shown below,

```xml
<endpoint name="FoodEP" xmlns="http://ws.apache.org/ns/synapse">
    <http method="get" uri-template="http://localhost:9192/service/foodservice">
        <authentication>
            <oauth>
                <passwordCredentials>
                    <clientId>clientId</clientId>
                    <clientSecret>clientSecret</clientSecret>
                    <username>internal-user</username>
                    <password>abc@123</password>
                    <tokenUrl>oauthServerUrl</tokenUrl>
                    <requestParameters>
                        <parameter name="account_id">{json-eval($.account_id)}</parameter>
                        <parameter name="user_role">tester</parameter>
                    </requestParameters>
                </passwordCredentials>
            </oauth>
        </authentication>
    </http>
</endpoint>
```

Fixes https://github.com/wso2/product-ei/issues/5492

### CODE REVIEW

File: modules/core/src/main/java/org/apache/synapse/config/xml/endpoints/HTTPEndpointSerializer.java [hunk 1]
```diff
@@ -107,18 +108,19 @@ private OMElement serializeOAuthConfiguration(OAuthConfiguredHTTPEndpoint oAuthC
...
             oauth.addChild(serializeAuthorizationCodeConfigurationElements(
                     (AuthorizationCodeHandler) oAuthConfiguredHTTPEndpoint.getOauthHandler()));
-        }
-
-        if (oAuthConfiguredHTTPEndpoint.getOauthHandler() instanceof ClientCredentialsHandler) {
+        } else if (oAuthConfiguredHTTPEndpoint.getOauthHandler() instanceof ClientCredentialsHandler) {
             oauth.addChild(serializeClientCredentialsConfigurationElements(
                     (ClientCredentialsHandler) oAuthConfiguredHTTPEndpoint.getOauthHandler()));
+        } else if (oAuthConfiguredHTTPEndpoint.getOauthHandler() instanceof PasswordCredentialsHandler) {
+            oauth.addChild(serializePasswordCredentialsConfigurationElements(
```
[the remark this answers is not in the document]
Comment (SanojPunchihewa) [reply modules/core/src/main/java/org/apache/synapse/config/xml/endpoints/HTTPEndpointSerializer.java]: Fixed with 8ae6b68

File: modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java [hunk 2]
```diff
@@ -115,11 +119,31 @@ private static OAuthHandler getSpecificOAuthHandler(OMElement oauthElement) {
         if (clientCredentialsElement != null) {
             oAuthHandler = getClientCredentialsHandler(clientCredentialsElement);
         }
+
+        if (passwordCredentialsElement != null) {
+            oAuthHandler = getPasswordCredentialsHandler(passwordCredentialsElement);
```
Comment (Nirothipan) [inline modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java:124]: can we re organize this if conditions, oAuthHandler will be updated only once

Comment (SanojPunchihewa) [reply modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java:120]: Fixed with d4bb790

File: modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java [hunk 3]
```diff
@@ -180,7 +196,38 @@ private static ClientCredentialsHandler getClientCredentialsHandler(
...
+            OMElement passwordCredentialsElement) {
+
+        String clientId = getChildValue(passwordCredentialsElement, OAuthConstants.OAUTH_CLIENT_ID);
+        String clientSecret = getChildValue(passwordCredentialsElement, OAuthConstants.OAUTH_CLIENT_SECRET);
+        String username = getChildValue(passwordCredentialsElement, OAuthConstants.OAUTH_USERNAME);
+        String password = getChildValue(passwordCredentialsElement, OAuthConstants.OAUTH_PASSWORD);
+        String tokenApiUrl = getChildValue(passwordCredentialsElement, OAuthConstants.TOKEN_API_URL);
+
+        if (username == null || password == null || tokenApiUrl == null || clientId == null || clientSecret == null) {
+            log.debug("Invalid PasswordCredentials configuration");
```
Comment (Nirothipan) [inline modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java:214]: shall we make is as error log ?

File: modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java [hunk 4]
```diff
@@ -132,9 +152,7 @@ private static AuthorizationCodeHandler getAuthorizationCodeHandler(OMElement au
         String tokenApiUrl = getChildValue(authCodeElement, OAuthConstants.TOKEN_API_URL);
 
         if (clientId == null || clientSecret == null || refreshToken == null || tokenApiUrl == null) {
-            if (log.isDebugEnabled()) {
-                log.error("Invalid AuthorizationCode configuration");
-            }
+            log.debug("Invalid AuthorizationCode configuration");
```
Comment (Nirothipan) [inline modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java:155]: same here

**Review by Nirothipan — DISMISSED** [review]

File: modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java (hunk 3 above)
Comment (SanojPunchihewa) [reply modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java]: Fixed with c5e1282

File: modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java (hunk 4 above)
Comment (SanojPunchihewa) [reply modules/core/src/main/java/org/apache/synapse/endpoints/oauth/OAuthUtils.java]: Fixed with c5e1282

**Review by Nirothipan — APPROVED** [review]
We need to update the docs with new configs

**Pull request merged**